tempesta_fw/classifier/frang.c

```diff
@@ -482,18 +484,66 @@ frang_http_host_check(const TfwHttpReq *req, FrangAcc *ra)
 	 * have a good regex library.
 	 * For now just linearize the Host header field TfwStr{} string.
 	 */
-	len = field.len + 1;
-	if ((buf = tfw_pool_alloc(req->pool, len)) == NULL)
-		return TFW_BLOCK;
-	tfw_str_to_cstr(&field, buf, len);
-	if (!tfw_addr_pton(buf, &addr)) {
+
+	urlhostlen = req->host.len;
+	hdrhostlen = field.len;
+
+	if (urlhostlen) {
+		if ((urlhost = tfw_pool_alloc(req->pool, urlhostlen + 1)) == NULL) {
+			TFW_ERR("Can not allocate memory\n");
+			ret = TFW_BLOCK;
+			goto finish;
+		}
+		tfw_str_to_cstr(&req->host, urlhost, urlhostlen + 1);
+		TFW_DBG2("req->host = [%s]\n", urlhost);
+	}
+
+	if ((hdrhost = tfw_pool_alloc(req->pool, hdrhostlen + 1)) == NULL) {
+		TFW_ERR("Can not allocate memory\n");
+		ret = TFW_BLOCK;
+		goto finish;
+	}
+
+	tfw_str_to_cstr(&field, hdrhost, hdrhostlen + 1);
+	if (!tfw_addr_pton(hdrhost, &addr)) {
 		frang_msg("Host header field contains IP address",
 			  &ra->addr, "\n");
-		tfw_pool_free(req->pool, buf, len);
-		return TFW_BLOCK;
+		ret = TFW_BLOCK;
+		goto finish;
 	}
-	tfw_pool_free(req->pool, buf, len);
-	return TFW_PASS;
+
+	if (req->uri_full && urlhostlen == 0 && hdrhostlen != 0) {
+		frang_msg("Host field not equal to host in URI",
+			  &ra->addr, "\n");
+		ret = TFW_BLOCK;
+		goto finish;
+	}
+
+	if (urlhostlen) {
+		if (hdrhostlen) {
+			if (hdrhostlen != urlhostlen || strcmp(hdrhost, urlhost) != 0) {
+				frang_msg("Host header is not equal to host in URL",
+					&ra->addr, "\n");
+				ret = TFW_BLOCK;
+				goto finish;
+			}
+		} else {
+			frang_msg("Host header is empty",
+			  &ra->addr, "\n");
+			ret = TFW_BLOCK;
+			goto finish;
+		}
+	} else {
+		TFW_DBG("No host in url\n");
+	}
```

Comment:
* Initialize `int ret = TFW_BLOCK` instead of `TFW_PASS` at function start.
* Remove all `ret = TFW_BLOCK` above, leaving just `goto finish;`
* Put `ret = TFW_PASS` here.